post.md

```diff
@@ -0,0 +1,427 @@
+
+```tsx
+"use client";
+
+import { deleteSomething } from "@/app/server-actions";
+import { ServerActionButton } from "./ServerActionButton";
+import { FormEvent } from "react";
+
+type Props = { id: string };
+
+export const ConfirmV2 = ({ id }: Props) => {
+  const onSubmit = (e: FormEvent<HTMLFormElement>) => {
+    if (!confirm("Are you sure?")) {
+      e.preventDefault();
+    }
+  };
+
+  return (
+    <form action={deleteSomething} onSubmit={onSubmit} className="inline-block">
+      <input type="hidden" name="id" value={id} />
+      <ServerActionButton />
+    </form>
+  );
+};
+```
+
+When the user clicks "OK" to confirm the submission, the `onSubmit` handler just lets the form do its thing and send off the data. But if the user clicks "Cancel", then the `onSubmit` handler uses `preventDefault` stopping the normal form behavior of submitting the data.
+
+## Custom modal dialog
+
+Sometimes we require more control over the design of the confirmation dialog. In this case we can create a custom modal dialog component and use it in place of the JavaScript `confirm` dialog.
+
+We need a `Modal` component that can be opened and closed and we need to pass the entire form to the modal as a child. This allows us to open the modal from the parent component and display the confirmation dialog to the user, once they confirm the action we can submit the form.
+
+You can reference the sample application for the `Modal` implementation, it exports both the component and a use hook `useModal` which exposes the `isOpen` state along with the `open` and `close` functions. Here is the parent component:
+
+```tsx
+"use client";
+
+import { deleteSomething } from "@/app/server-actions";
+import { ServerActionButton } from "./ServerActionButton";
+import { Modal, useModal } from "./Modal";
+import { Button } from "./Button";
+
+type Props = { id: string };
+
+export const DialogConfirm = ({ id }: Props) => {
+  const { isOpen, open, close } = useModal();
+
+  return (
+    <>
+      <div className="flex flex-col h-full text-center">
+        <div className="bg-white py-8 px-8 dark:bg-slate-900">
+          <Button text="Click Me!" onClick={open} />
+        </div>
+      </div>
+
+      <Modal
+        isOpen={isOpen}
+        close={close}
+        title="Confirm"
+        content="Are you sure?"
+      >
+        <form action={deleteSomething} className="inline-block">
+          <input type="hidden" name="id" value={id} />
+
+          <Button text="Cancel" onClick={close} className="mr-2" />
+          <ServerActionButton text="Ok" />
+        </form>
+      </Modal>
+    </>
+  );
+};
+```
+
+This works great! You click the 'submit button' and the modal dialog opens, you confirm the action by clicking the real submit button or cancel the action by clicking the cancel button closing the modal.
```

Comment:
The sentence structure is a little confusing for me. Maybe try something like this, "This works great! When a user clicks the submit button, the modal dialog opens up. The user can then confirm the action by clicking the 'real' submit button, or cancel the action by clicking the cancel button (and thereby closing the modal).